### PR TITLE
prometheus-operator config update

### PR DESCRIPTION
**Pull request by knrt10**
Add user-configurable tolerations for prometheus operator and it's
components.

Change configuration for alertmanager. Now it is configured in a
different block

closes: #1527 
Signed-off-by: knrt10 <kautilya@kinvolk.io>

### Release notes

Alert manager is not configured as a different block

```hcl
# old
alertmanager_retention    = "360h"
alertmanager_external_url = "https://api.example.com/alertmanager"
alertmanager_config       = file("alertmanager-config.yaml")
alertmanager_node_selector = {
  "kubernetes.io/hostname" = "worker3"
}

# new
alertmanager {
  retention    = "360h"
  external_url = "https://api.example.com/alertmanager"
  config       = file("alertmanager-config.yaml")
  node_selector = {
    "kubernetes.io/hostname" = "worker3"
  }
}
```

Operator is also now configured as a new block

```hcl
# old
prometheus_operator_node_selector = {
  "kubernetes.io/hostname" = "worker3"
}

# new
operator {
  node_selector = {
    "kubernetes.io/hostname" = "worker3"
  }
}
```

### CODE REVIEW

File: pkg/components/prometheus-operator/component.go
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:82]: Perhaps we should have a struct for `PrometheusOperator` fields.

File: pkg/components/prometheus-operator/component.go
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:92]: Ditto AlertManager

File: pkg/components/prometheus-operator/component.go
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:214]:
```suggestion
		return nil, fmt.Errorf("rendering prometheus tolerations: %w", err)
```

File: pkg/components/prometheus-operator/template.go
Comment (invidian) [inline pkg/components/prometheus-operator/template.go:94]: I feel like we could get rid of this condition.

**Review by invidian — CHANGES_REQUESTED** [review]
Perhaps we could add some tests.

File: docs/configuration-reference/components/prometheus-operator.md
Comment (invidian) [inline docs/configuration-reference/components/prometheus-operator.md:50]: I wonder if we could somehow improve this experience for users. E.g. being able to identify what workloads are part of the component and then do some smart mapping to worker groups, so you don't have to specify those manually.

**Comment by knrt10** [comment]
@invidian do we want package tests or e2e test?

File: pkg/components/prometheus-operator/component.go
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:75]: Can introduce new block in separate commit from adding new fields? That will make reviewing easier.

**Review by invidian — COMMENTED** [review]
>@invidian do we want package tests or e2e test?

I think having at least manifest rendering tests touching all new options would be good.

File: pkg/components/prometheus-operator/component.go
Comment (invidian) [inline pkg/components/prometheus-operator/component.go:147]: We shouldn't need an empty object here.

**Review by invidian — COMMENTED** [review]
Just one more nit, otherwise looks good!

File: pkg/components/prometheus-operator/component.go
Comment (knrt10) [reply pkg/components/prometheus-operator/component.go:147]: If not done it leads to invalid memory address or nil pointer dereference. It should not though, I don't know why this happens.

**Comment by invidian** [comment]
I suggest adding:
```patch
diff --git a/pkg/components/prometheus-operator/component_test.go b/pkg/components/prometheus-operator/component_test.go
index 93c2911a..c161b7d9 100644
--- a/pkg/components/prometheus-operator/component_test.go
+++ b/pkg/components/prometheus-operator/component_test.go
@@ -24,6 +24,8 @@ import (

 //nolint:funlen
 func TestRenderManifest(t *testing.T) {
+       t.Parallel()
+
        tests := []struct {
                desc    string
                hcl     string
@@ -143,6 +145,8 @@ component "prometheus-operator" {
        for _, tc := range tests {
                tc := tc
                t.Run(tc.desc, func(t *testing.T) {
+                       t.Parallel()
+
                        b, d := util.GetComponentBody(tc.hcl, Name)
                        if d != nil {
                                t.Fatalf("error getting component body: %v", d)
@@ -177,6 +181,8 @@ component "prometheus-operator" {

 //nolint:funlen
 func TestConversion(t *testing.T) {
+       t.Parallel()
+
        testCases := []struct {
                name                 string
                inputConfig          string
```

This cuts down test execution time from 14 seconds to 4 seconds on my machine. Still slow, but better.

File: pkg/components/prometheus-operator/template.go
Comment (invidian) [inline pkg/components/prometheus-operator/template.go:98]: This has been removed a long time ago from upstream: https://github.com/prometheus-community/helm-charts/blob/7fcf3417466f3a3c6c1388167b156b0806a7eb96/charts/kube-prometheus-stack/README.md#from-10x-to-11x

Can we remove it?

File: pkg/components/prometheus-operator/template.go
Comment (invidian) [inline pkg/components/prometheus-operator/template.go:103]: If both webhook is disabled and tolerations for it are set, there will be duplicated key in YAML.

File: pkg/components/prometheus-operator/component.go
Comment (invidian) [reply pkg/components/prometheus-operator/component.go:147]: I suggest the following patch:
```patch
diff --git a/pkg/components/prometheus-operator/component.go b/pkg/components/prometheus-operator/component.go
index 4aa47328..d648eb24 100644
--- a/pkg/components/prometheus-operator/component.go
+++ b/pkg/components/prometheus-operator/component.go
@@ -144,7 +144,6 @@ func NewConfig() *component {
                        StorageSize: "50Gi",
                },
                Namespace: "monitoring",
-               Operator:  &Operator{},
                Monitor: &Monitor{
                        Etcd:                  true,
                        KubeControllerManager: true,
@@ -226,14 +225,16 @@ func (c *component) RenderManifests() (map[string]string, error) {
                return nil, fmt.Errorf("rendering prometheus tolerations: %w", err)
        }

-       c.Operator.TolerationsRaw, err = util.RenderTolerations(c.Operator.Tolerations)
-       if err != nil {
-               return nil, fmt.Errorf("rendering operator tolerations: %w", err)
-       }
+       if c.Operator != nil {
+               c.Operator.TolerationsRaw, err = util.RenderTolerations(c.Operator.Tolerations)
+               if err != nil {
+                       return nil, fmt.Errorf("rendering operator tolerations: %w", err)
+               }

-       c.Operator.AdmissionWebhookTolerationsRaw, err = util.RenderTolerations(c.Operator.AdmissionWebhookTolerations) //nolint:lll
-       if err != nil {
-               return nil, fmt.Errorf("rendering operator admission webhook tolerations: %w", err)
+               c.Operator.AdmissionWebhookTolerationsRaw, err = util.RenderTolerations(c.Operator.AdmissionWebhookTolerations) //nolint:lll
+               if err != nil {
+                       return nil, fmt.Errorf("rendering operator admission webhook tolerations: %w", err)
+               }
        }

        c.AlertManager.TolerationsRaw, err = util.RenderTolerations(c.AlertManager.Tolerations)
diff --git a/pkg/components/prometheus-operator/template.go b/pkg/components/prometheus-operator/template.go
index 7b9d93f6..9d3d08d6 100644
--- a/pkg/components/prometheus-operator/template.go
+++ b/pkg/components/prometheus-operator/template.go
@@ -91,29 +91,24 @@ kubeEtcd:
   enabled: {{.Monitor.Etcd}}
 prometheus-node-exporter:
   service: {}
-{{ if (or .Operator.NodeSelector .DisableWebhooks .Operator.Tolerations ) }}
 prometheusOperator:
-  {{- if .DisableWebhooks }}
-  tlsProxy:
-    enabled: false
-  admissionWebhooks:
-    enabled: false
-  {{- end }}
-  {{- if .Operator.AdmissionWebhookTolerations }}
   admissionWebhooks:
+    enabled: {{ not .DisableWebhooks  }}
+  {{- with .Operator }}
+  {{- if .AdmissionWebhookTolerations }}
     patch:
-      tolerations: {{ .Operator.AdmissionWebhookTolerationsRaw }}
+      tolerations: {{ .AdmissionWebhookTolerationsRaw }}
   {{- end }}
-  {{- if .Operator.NodeSelector }}
+  {{- if .NodeSelector }}
   nodeSelector:
-    {{ range $key, $value := .Operator.NodeSelector }}
+    {{ range $key, $value := .NodeSelector }}
     {{ $key }}: {{ $value }}
     {{ end }}
   {{- end }}
-  {{- if .Operator.Tolerations }}
-  tolerations: {{ .Operator.TolerationsRaw }}
+  {{- if .Tolerations }}
+  tolerations: {{ .TolerationsRaw }}
+  {{- end }}
   {{- end }}
-{{ end }}
 prometheus:
   {{ if .Prometheus.Ingress }}
   ingress:
```

**Review by invidian — COMMENTED** [review]
Looks OK

**Comment by invidian** [comment]
@knrt10 changes looks good to me, but I think some other maintainer should approve it for merging.

**Comment by knrt10** [comment]
@invidian I think you said Looks OK and submitted it via request changes. So had to request again 😄

File: pkg/components/prometheus-operator/component_test.go
Comment (surajssd) [inline pkg/components/prometheus-operator/component_test.go:115]: How about a conversion test to test that these values are getting on to the actual config at the Kubernetes level.

File: pkg/components/prometheus-operator/component.go
Comment (surajssd) [inline pkg/components/prometheus-operator/component.go:57]: The second commit (3c1689f5b921dd87b25b5e367178c8e1017fb258) says `Unverified`.

File: pkg/components/prometheus-operator/component_test.go
Comment (surajssd) [reply pkg/components/prometheus-operator/component_test.go:115]: Does not have to be a part of this function. It could be a standalone test to verify all the changes that went in this PR.

**Review by surajssd — CHANGES_REQUESTED** [review]

File: pkg/components/prometheus-operator/component.go
Comment (invidian) [reply pkg/components/prometheus-operator/component.go:57]: This is fine. @knrt10 does not sign it's commits and my account requires a signature, as I always sign.

File: pkg/components/prometheus-operator/component_test.go
Comment (invidian) [reply pkg/components/prometheus-operator/component_test.go:115]: @surajssd Can you please link some examples of such tests to make it easier for @knrt10 to implement them?

Comment (knrt10) [reply pkg/components/prometheus-operator/component_test.go:115]: I think I understand what Suraj is trying to say. I will implement it

File: pkg/components/prometheus-operator/component.go
Comment (knrt10) [reply pkg/components/prometheus-operator/component.go:57]: Signed now. On next push it should be verified I guess

**Review by surajssd — DISMISSED** [review]

**Comment by surajssd** [comment]
@knrt10 can you rebase on the current master the ARM issue will be solved.

**Comment by knrt10** [comment]
AWS is still failing on certificate rotation part.

**Review by surajssd — APPROVED** [review]

**Comment by knrt10** [comment]
finally, yes

**Pull request merged**
